json/test_yajl.c

```diff
@@ -0,0 +1,211 @@
+	context_t* ctx = ctx_v;
+
+	if (ctx->depth == 1) {
+		ctx->in_coord = strncmp("coordinates", (const char*)key, length) == 0;
+	} else if (ctx->depth == 2 && ctx->in_coord) {
+		ctx->item = (length > 0 ? key[0] : '\0');
+	}
+
+	return 1;
+}
+
+int cb_end_map(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	ctx->depth--;
+
+	return 1;
+}
+
+int cb_end_array(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	if (ctx->depth == 1 && ctx->in_coord) {
+		ctx->value.x /= ctx->len;
+		ctx->value.y /= ctx->len;
+		ctx->value.z /= ctx->len;
+	}
+
+	return 1;
+}
+
+static yajl_callbacks ycallbacks = {
+	NULL, /* null */
+	NULL, /* boolean */
+	NULL, /* integer */
+	&cb_double,
+	NULL, /* number */
+	NULL, /* string */
+	&cb_start_map,
+	&cb_map_key,
+	&cb_end_map,
+	NULL, /* start_array */
+	&cb_end_array
+};
+
+context_t init_context() {
+	return (context_t){
+		.depth = 0,
+		.len = 0,
+		.item = '\0',
+		.in_coord = false,
+		.value = {
+			.x = 0.0,
+			.y = 0.0,
+			.z = 0.0
+		}
+	};
+}
+
+int verify(const char* buf, size_t size, const coordinate_t* expected) {
+	context_t ctx = init_context();
+
+	yajl_handle yajl = yajl_alloc(&ycallbacks, NULL, &ctx);
+
+	yajl_parse(yajl, (const unsigned char*)buf, size);
+	yajl_complete_parse(yajl);
+
+	yajl_free(yajl);
+
+	if (ctx.value.x != expected->x || ctx.value.y != expected->y || ctx.value.z != expected->z) {
+		fprintf(stderr, "{%lf, %lf, %lf} != {%lf, %lf, %lf}\n", ctx.value.x, ctx.value.y, ctx.value.z, expected->x, expected->y, expected->z);
+
+		return 1;
+	}
+
+	return 0;
+}
+
+int main() {
+	const char test_buffer_1[] = "{\"coordinates\":[{\"x\":2.0,\"y\":0.5,\"z\":0.25}]}";
+	const char test_buffer_2[] = "{\"coordinates\":[{\"y\":0.5,\"x\":2.0,\"z\":0.25}]}";
+
+	const coordinate_t expected = {
+		.x = 2.0,
+		.y = 0.5,
+		.z = 0.25
+	};
+
+	if (verify(test_buffer_1, sizeof(test_buffer_1) / sizeof(test_buffer_1[0]), &expected)) {
+		exit(EXIT_FAILURE);
+	}
+
+	if (verify(test_buffer_2, sizeof(test_buffer_2) / sizeof(test_buffer_2[0]), &expected)) {
+		exit(EXIT_FAILURE);
+	}
+
+	struct stat file_stat;
+	int ret = 0;
+	unsigned char* buf = NULL;
+
+	int fd = open("/tmp/1.json", O_RDONLY);
+	if (fd < 0) {
+		fprintf(stderr, "Cannot open /tmp/1.json: %s\n", strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+
+	ret = fstat(fd, &file_stat);
+	if (ret < 0) {
+		fprintf(stderr, "Cannot fstat /tmp/1.json: %s\n", strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+
+	buf = calloc(file_stat.st_size, sizeof(unsigned char));
+	if (!buf) {
+		fprintf(stderr, "Cannot allocate %ld bytes memory\n", file_stat.st_size);
+		exit(EXIT_FAILURE);
+	}
+
+	context_t ctx = init_context();
+	yajl_handle yajl = yajl_alloc(&ycallbacks, NULL, &ctx);
+
+	notify_with_pid("C/" COMPILER " (yajl)");
+
+	ret = read(fd, buf, file_stat.st_size);
+	if (ret != file_stat.st_size) {
+		fprintf(stderr, "Cannot read /tmp/1.json: %s\n", strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+
+	yajl_parse(yajl, buf, file_stat.st_size);
```

Comment:
A little bit abstraction leak here (file size), and I guess it would be nice to separate file operations and parsing by introduce an auxiliary struct:

```c
struct str_s {
  char *buf;
  size_t buf_sz;
};
typedef struct str_s str_t;

void str_free(str_t* s) { free(s->buf); }

str_t readfile(const char* filename) {...}

int main() {
  ...
  __attribute__((__cleanup__(str_free))) str_t str = readfile("/tmp/1.json");
  ...
  yajl_parse(yajl, str.buf, str.buf_sz);
}

```